[HLSL] Appropriately set function attribute optnone

bharadwajy opened the pull request:
When optimization is disabled, set `optnone` attribute all module entry functions.

Updated test in accordance with the change.

With this change, the plan is to eliminate generation of non-standard DXIL metadata `dx.disable_optimizations` and use the function attribute to detect setting of shader flag `DisableOptimizations` in a follow-on PR [with these changes](https://github.com/llvm/llvm-project/compare/main...bharadwajy:llvm-project:delete-disable-opt-metadata?expand=1) to address issues #124796 and #112263.

Closes #124796

Commit:
[HLSL] Set function optnone attribute appropriately

When optimization is disabled, set optnone attribute
  - for all module functions when targetting Library shaders
  - only for entry function when targetting non-Library shaders

Update tests in accordance with the change.

bharadwajy commented:
Need to verify a couple more function attribute combinations. Changed it to draft mode till that is done. 

Please hold off reviews.

bharadwajy commented:
> Need to verify a couple more function attribute combinations. Changed it to draft mode till that is done.
> 
> Please hold off reviews.

PR ready for review. Thanks!

bogner commented on clang/lib/CodeGen/CGHLSLRuntime.cpp:
```diff
@@ -345,6 +345,9 @@ void clang::CodeGen::CGHLSLRuntime::setHLSLEntryAttributes(
+  if (CGM.getCodeGenOpts().OptimizationLevel == 0) {
+    Fn->addFnAttr(llvm::Attribute::OptimizeNone);
+  }
   Fn->addFnAttr(llvm::Attribute::NoInline);
```
Given that these already have "noinline", I'm surprised that the logic in "SetLLVMFunctionAttributesForDefinition" doesn't already put optnone on these functions. Is something undoing this later?

bogner replied:
Ok. I think a comment to the effect of "We need to manually set attributes here instead of relying on `SetLLVMFunctionAttributesForDefinition` to pick them up since these functions are injected by the compiler and won't go through the normal flow" (please reword as necessary to be accurate...) would be a good idea here.

bharadwajy replied:
> Ok. I think a comment to the effect of "We need to manually set attributes here instead of relying on `SetLLVMFunctionAttributesForDefinition` to pick them up since these functions are injected by the compiler and won't go through the normal flow" (please reword as necessary to be accurate...) would be a good idea here.

Comment added. Thanks!

bogner commented on clang/lib/CodeGen/CGHLSLRuntime.cpp:
```diff
@@ -446,6 +449,13 @@ void CGHLSLRuntime::setHLSLFunctionAttributes(const FunctionDecl *FD,
     const StringRef ExportAttrKindStr = "hlsl.export";
     Fn->addFnAttr(ExportAttrKindStr);
   }
+  llvm::Triple T(Fn->getParent()->getTargetTriple());
+  if (T.getEnvironment() == llvm::Triple::EnvironmentType::Library) {
+    if (CGM.getCodeGenOpts().OptimizationLevel == 0) {
+      Fn->addFnAttr(llvm::Attribute::OptimizeNone);
+      Fn->addFnAttr(llvm::Attribute::NoInline);
+    }
+  }
```
Do we want to do this on all functions in a library or just entry points and exported functions? In any case, it really would be preferable if "SetLLVMFunctionAttributesForDefinition" did the right thing (whatever that may be) rather than us needing to duplicate that logic here...

Commit:
[HLSL] Delete the change that sets non-entry functions of a library
shader as optnone if optimization is disabled.

Update corresponding test changes

bogner reviewed (approved):
This looks reasonable to me. It would be good to get some feedback from @pow2clk and @llvm-beanz since they've done a bunch of the work in this area.

Commit:
Add comment

llvm-beanz commented on clang/lib/CodeGen/CGHLSLRuntime.cpp:
```diff
@@ -345,6 +345,14 @@ void clang::CodeGen::CGHLSLRuntime::setHLSLEntryAttributes(
+  // later in the compiler-flow for such module functions is not aware of and
+  // hence not able to set attributes of the newly materialized entry functions.
+  // So, set attributes of entry function here, as appropriate.
+  if (CGM.getCodeGenOpts().OptimizationLevel == 0) {
+    Fn->addFnAttr(llvm::Attribute::OptimizeNone);
+  }
```
nit: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements
```suggestion
  if (CGM.getCodeGenOpts().OptimizationLevel == 0)
    Fn->addFnAttr(llvm::Attribute::OptimizeNone);
```

llvm-beanz reviewed (approved):
One nit, otherwise LGTM.

Commit:
Delete braces around simple single-statement body of if statement
